### PR TITLE
plantuml-server 1.2021.12 -> 1.2022.2

### PR DESCRIPTION
**Pull request by nazarewk**
###### Description of changes

Suplements https://github.com/NixOS/nixpkgs/pull/163416 in updating `plantuml` suite to consistent versions.

There were following issues resolved on the way:
- `plantuml-server` required 11+ version of JDK, updated the module to keep consistent versions,
- maven was not configured for reproducible builds and embedded timestamps of files in `*.war` output package
- `plantuml-server` started requiring Jetty 11+ (due to `javax` being renamed to `jakarta` [for legal reasons](https://dzone.com/articles/from-javax-to-jakarta-a-simple-proof-of-concept))

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`):
	1. I've reconstructed the `services.plantuml-server` command from this branch results:
		  ```
		  nix build '.#plantuml-server' '.#jdk' '.#jetty'
		  "${PWD}/result-1/bin/java" -jar "${PWD}/result-2/start.jar" --module=deploy,http,jsp jetty.home="${PWD}/result-2" jetty.base="${PWD}/result" jetty.http.host=127.0.0.1 jetty.http.port=7890
		  ```
	2. Confirmed it's working by going to `http://127.0.0.1:7890/plantuml` in the browser
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by drupol — APPROVED** [review]

**Comment by nazarewk** [comment]
There is some weird error on my local build:
```
plantuml-server> [INFO] --- maven-compiler-plugin:3.8.1:compile (default-compile) @ plantumlservlet ---
plantuml-server> [INFO] Changes detected - recompiling the module!
plantuml-server> [INFO] Compiling 18 source files to /build/source/target/classes
plantuml-server> [INFO] ------------------------------------------------------------------------
plantuml-server> [INFO] BUILD FAILURE
plantuml-server> [INFO] ------------------------------------------------------------------------
plantuml-server> [INFO] Total time:  1.490 s
plantuml-server> [INFO] Finished at: 2022-03-09T11:30:36Z
plantuml-server> [INFO] ------------------------------------------------------------------------
plantuml-server> [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project plantumlservlet: Fatal error compiling: invalid target release: 11 -> [Help 1]
plantuml-server> [ERROR]
plantuml-server> [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
plantuml-server> [ERROR] Re-run Maven using the -X switch to enable full debug logging.
plantuml-server> [ERROR]
plantuml-server> [ERROR] For more information about the errors and possible solutions, please read the following articles:
plantuml-server> [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
plantuml-server> timeout, restart maven to continue downloading

```

**Review by truh — APPROVED** [review]

**Comment by truh** [comment]
@nazarewk I encountered this problem as well when I tried to update this package. My workaround was to use a different jdk `jdk17_headless` https://github.com/NixOS/nixpkgs/pull/159063. I didn't go through with it since the output hash of deps kept changing.

**Comment by drupol** [comment]
How about changing in the `nativeBuildInputs`:  `jdk8_headless` in `jdk17_headless` ? Can you give it a try ?

Also, is there any kind of documentation that says what is the required version of the JDK for this?

**Comment by nazarewk** [comment]
> Also, is there any kind of documentation that says what is the required version of the JDK for this?

Looks like it requires JDK 11+ and Maven 3.0.2+ https://github.com/plantuml/plantuml-server/blob/0330b7c4a20292082465c021ec415cc013167998/README.md?plain=1#L27-L30

**Comment by nazarewk** [comment]
@truh @drupol looks like it works with `jdk17_headless`, but just like truh said, the output is not stable

**Comment by drupol** [comment]
Why not using `jdk11`?

**Comment by nazarewk** [comment]
> Why not using `jdk11`?

This is the same for `jdk11_headless`

**Comment by nazarewk** [comment]
> > > Why not using `jdk11`?
> > 
> > 
> > This is the same for `jdk11_headless`
> 
> Then maybe we should open an issue upstream ?

I mean the output is unstable for both `jdk11_headless` and `jdk17_headless`

**Comment by nazarewk** [comment]
Looks like timestamps in output `*.war` are differing:
nazarewk@nazarewk-krul /tmp/plantuml-server % ls -la            
total 38560
drwxr-xr-x  4 nazarewk users      120 Mar 10 11:39 .
drwxrwxrwt 27 root     root       780 Mar 10 11:38 ..
drwxr-xr-x  5 nazarewk users      200 Mar 10 11:40 plantuml-1
-r--r--r--  1 nazarewk users 19740424 Mar 10 11:34 plantuml-1.war
drwxr-xr-x  5 nazarewk users      200 Mar 10 11:40 plantuml-2
-r--r--r--  1 nazarewk users 19740424 Mar 10 11:36 plantuml-2.war
nazarewk@nazarewk-krul /tmp/plantuml-server % ls -la plantuml-1
total 20
drwxr-xr-x 5 nazarewk users  200 Mar 10 11:40 .
drwxr-xr-x 4 nazarewk users  120 Mar 10 11:39 ..
-rw-r--r-- 1 nazarewk users 1884 Mar 10 10:05 error.jsp
-rw-r--r-- 1 nazarewk users 1150 Mar 10 10:05 favicon.ico
-rw-r--r-- 1 nazarewk users  123 Mar 10 10:05 footer.jspf
-rw-r--r-- 1 nazarewk users 3870 Mar 10 10:05 index.jsp
drwxr-xr-x 3 nazarewk users   80 Mar 10 11:40 META-INF
-rw-r--r-- 1 nazarewk users  947 Mar 10 10:05 plantuml.css
drwxr-xr-x 2 nazarewk users  120 Mar 10 10:05 resource
drwxr-xr-x 4 nazarewk users  100 Mar 10 10:05 WEB-INF
nazarewk@nazarewk-krul /tmp/plantuml-server % ls -la plantuml-2
total 20
drwxr-xr-x 5 nazarewk users  200 Mar 10 11:40 .
drwxr-xr-x 4 nazarewk users  120 Mar 10 11:39 ..
-rw-r--r-- 1 nazarewk users 1884 Mar 10 10:36 error.jsp
-rw-r--r-- 1 nazarewk users 1150 Mar 10 10:36 favicon.ico
-rw-r--r-- 1 nazarewk users  123 Mar 10 10:36 footer.jspf
-rw-r--r-- 1 nazarewk users 3870 Mar 10 10:36 index.jsp
drwxr-xr-x 3 nazarewk users   80 Mar 10 11:40 META-INF
-rw-r--r-- 1 nazarewk users  947 Mar 10 10:36 plantuml.css
drwxr-xr-x 2 nazarewk users  120 Mar 10 10:36 resource
drwxr-xr-x 4 nazarewk users  100 Mar 10 10:36 WEB-INF
nazarewk@nazarewk-krul /tmp/plantuml-server % diff -r plantuml-* 
diff: extra operand 'plantuml-2'
diff: Try 'diff --help' for more information.
2 nazarewk@nazarewk-krul /tmp/plantuml-server % diff -r plantuml-{1,2}

**Comment by nazarewk** [comment]
Looks like there is `reproducible-build-maven-plugin`, but i have no clue whether it's possible or how to apply it to this use case https://zlika.github.io/reproducible-build-maven-plugin/

**Comment by nazarewk** [comment]
FYI: `-Dproject.build.outputTimestamp=0` to `mvn package` was enough to trigger reproducible builds as per https://maven.apache.org/guides/mini/guide-reproducible-builds.html#how-to-test-my-maven-build-reproducibility

**Comment by Artturin** [comment]
Has this been tested if so I will merge

**Comment by nazarewk** [comment]
looks like this actually doesn't work yet due to switching from `javax` to `jakarta` and Jetty not keeping up with it:
- https://plantuml.com/server
- https://webtide.com/renaming-from-javax-to-jakarta/

**Review by emmanuelrosa — APPROVED** [review]
There are three stable versions of Jetty: 9.4, 10, and 11.
Since there are no other packages in Nixpkgs which use Jetty, I'm OK with upgrading the Jetty package to version 11.

If other packages arise in the future needing version 9.4 or 10, I'll deal with it then.

**Review by drupol — APPROVED** [review]
LGTM

**Comment by truh** [comment]
I did a little bit of testing. When I started the service from this branch SVG generation didn't work. I was able to fix the problem by replacing `jdk17_headless` with `jdk` in `nixos/modules/services/web-apps/plantuml-server.nix`.

File: nixos/modules/services/web-apps/plantuml-server.nix [hunk 1]
```diff
@@ -20,6 +20,21 @@ in
         description = "PlantUML server package to use";
       };
 
+      packages = {
+        jdk = mkOption {
+          type = types.package;
+          default = pkgs.jdk17_headless;
+          defaultText = literalExpression "pkgs.jdk17_headless";
```
Comment (truh) [inline nixos/modules/services/web-apps/plantuml-server.nix:27]:
```suggestion
          default = pkgs.jdk;
          defaultText = literalExpression "pkgs.jdk";
```

**Comment by nazarewk** [comment]
I have tested it with `jdk` by reconstructing the commandline from `services.plantuml-server`, I'm on flake based system and those instructions above to spawn a VM gave me very weird results (it had GDM and hostname the same as my host system).

**Review by emmanuelrosa — APPROVED** [review]

**Review by truh — APPROVED** [review]

**Comment by vcunat** [comment]
This broke the build on Hydra, I think: https://hydra.nixos.org/eval/1748939?filter=plantuml#tabs-now-fail

**Comment by nazarewk** [comment]
> This broke the build on Hydra, I think: [hydra.nixos.org/eval/1748939?filter=plantuml#tabs-now-fail](https://hydra.nixos.org/eval/1748939?filter=plantuml#tabs-now-fail)

Any idea what is wrong? I have trouble pinpointing the issue (responsible logs or anything)

**Comment by vcunat** [comment]
Locally I'm getting
```
hash mismatch in fixed-output derivation '/nix/store/7wvrgia71mrr658rpxak3dw6aszqid0b-plantuml-server-1.2022.2-deps':
  wanted: sha256:1042nkxic732ppjma2lc894c8812h4js8dijm73k05v3gq2q45q2
  got:    sha256:1vc2j6xk1csagbsx15qdgxsglhmm75rsj8npjf19qj0i8k8jd46y
cannot build derivation '/nix/store/mllvsyizjl5frnwpsic5gp1msivxz0n5-plantuml-server-1.2022.2.drv': 1 dependencies couldn't be built
```

**Comment by nazarewk** [comment]
what system are you on?

**Comment by vcunat** [comment]
x86_64 NixOS.  This build was `nix build -f. plantuml-server` on commit c0ede2616 specifically.

**Comment by nazarewk** [comment]
It works just fine for me on the same system using `nix build '.#plantuml-server'` and `fad085c08f97ff526cda9bcfb83c6c00291c66fd`

**Comment by truh** [comment]
@nazarewk did it rebuild the deps for you or were the deps still in your nix store?

**Comment by nazarewk** [comment]
I've `sha256-3pAm0UQRSJyCk9ciqXM5tUL6dH8Nl9D1ekqzMLuRgu0=` on 2 conscutive rebuild with `nix-collect-garbage -d` in between, which is not what was in the PR.

Might be that dependencies are not pinned in `plantuml-server` and get different results in different times?

**Comment by truh** [comment]
I think it's likely that the maven artifacts in the repository are mutable.

**Comment by truh** [comment]
Maybe it would be better to just download the war file from the Github release and call it a day.

**Pull request merged**